app/models/token.rb

```diff
@@ -21,21 +21,28 @@ def payload
                  end
   rescue JWT::ExpiredSignature => e
     Rails.logger.info(e.message, token: @token_string)
-    raise error_klass("Validation error: #{e.message}")
+    raise error_klass(e.message)
+  rescue JWT::DecodeError => e
+    raise error_klass(e.message)
   end
 
   def public_key
     decoded_token = JWT.decode(@token_string, nil, false, algorithm: 'RS256')
     kid = decoded_token[1]['kid']
-    OIDC::KeyService.get_key(kid)
+    key = OIDC::KeyService.get_key(kid)
+    if key.blank?
+      Rails.logger.info('Public key not found', kid: kid, exp: decoded_token[0]['exp'])
+      raise error_klass("Public key not found for kid specified in token: '#{kid}'")
+    end
```

Comment:
Since this function will never return a blank key, I think the `return` on line 18 can go away.

Reply:
Good point!